common/Obelisk/OAuth/State/Internal.hs

```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+{-| Internals of "Obelisk.OAuth.State". Use if you want to provide your own
+   `OAuthState` generator, but make sure you are using truly random values!
+-}
+module Obelisk.OAuth.State.Internal where
+
+import           Control.Monad.IO.Class (MonadIO, liftIO)
+import           Data.Aeson             (FromJSON, ToJSON)
+import qualified Data.ByteString.Base16 as Base16
+import           Data.Text              (Text)
+import qualified Data.Text.Encoding     as T
+import           GHC.Generics           (Generic)
+import           System.Entropy         (getEntropy, getHardwareEntropy)
+
+
+-- | Oauth state to prevent CSRF attacks.
+newtype OAuthState = OAuthState { unOAuthState :: Text }
+  deriving (Generic, Show, Read, Eq, Ord, ToJSON, FromJSON)
+
+-- | Make an `OAuthState` from some Text.
+--
+--   If you need a brand new `OAuthState` you should use `genOAuthState`. This
+--   function is useful for deserialization of server responses for example.
+unsafeMkOAuthState :: Text -> OAuthState
+unsafeMkOAuthState = OAuthState
```

Comment:
What makes this unsafe?  Does OAuthState enforce any requirements on its data?

Reply:
Yep. The state is there to prevent CSRF attacks - but this protection only works if the given state is not guessable. so `OAuthState "foobar"` - would be a bad idea.